Update CI/2FA docs (#779)
fastlane/docs: docs/best-practices/continuous-integration.md

_fastlane_ works very well in Continuous Integration setups. It for example automatically generates a JUnit report for you that allows Continuous Integration systems, like `Jenkins`, access the results of your deployment.

> - [Intergrations](#integrations)

Reviewer: ```suggestion
- [Integrations](#integrations)
```

## Two-step or Two-factor auth

> Fastlane fully support 2 step verification and 2 Factor Authentication for logging in to your Apple ID and Apple Developer account.

Reviewer: ```suggestion
_fastlane_ fully supports 2-step verification and 2-factor authentication for logging in to your Apple ID and Apple Developer account. 🌟
```

The easiest way to get _fastlane_ running on a CI system is to create a separate Apple ID that doesn't have 2-factor auth enabled, with a long, randomly generated password. Additionally make sure the newly created Apple account has limited access to only the apps and resources it needs.

Creating a separate Apple ID allows you to limit the permission scope, have a randomly generated password, and will make it much easier for you to set up CI using _fastlane_.

### Security code and session

> When your Apple account has 2 Factor Authentication (or 2 step verification) enabled, you will be asked to verify your identity by entering a security code. If you already have a trusted device configured for your account, then the code will appear on the device. If you don't have any devices configured, but have trusted a phone number, then the code will be sent to your phone. The resulting session will be stored in `~/.fastlane/spaceship/[email]/cookie`. The session should be valid for about one month, however there is no way to test this without actually waiting for over a month.

Reviewer: ```suggestion
When your Apple account has 2-factor authentication (or 2-step verification) enabled, you will be asked to verify your identity by entering a security code. If you already have a trusted device configured for your account, then the code will appear on the device. If you don't have any devices configured, but have trusted a phone number, then the code will be sent to your phone. The resulting session will be stored in `~/.fastlane/spaceship/[email]/cookie`. The session should be valid for about one month, however there is no way to test this without actually waiting for over a month.
```

### Use of application specific passwords and `spaceauth`

When you can not enter this token, as on a Continuous Integration system, you have to use other ways to log in:

#### Application specific passwords

If you want to upload builds to TestFlight / App Store Connect from your CI machine, you need to generate an application specific password:

1. Visit [appleid.apple.com/account/manage](https://appleid.apple.com/account/manage)
1. Generate a new application specific password
1. Provide the application specific password using the environment variable `FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD`

#### `spaceauth`

> Because your CI machine will not be able to prompt you for your two-step or two-factor auth information, you also need to generate a login session for your CI machine in advance. You can get on your local machine this by running:

Reviewer: ```suggestion
Because your CI machine will not be able to prompt you for your 2-step or 2-factor auth information, you also need to generate a login session for your CI machine in advance. You can get this on your local machine by running:
```

```
fastlane spaceauth -u user@email.com
```

This will generate a token you can set using the `FASTLANE_SESSION` environment variable on your CI system.

Please note:

- An Apple ID session is only valid for a certain region, meaning if your CI system is in a different region than your local machine, you'll run into issues
- An Apple ID session is only valid for up to a month, meaning you'll have to generate a new session every month. Usually you'd only know about it when your build starts failing

There is nothing _fastlane_ can do better in that regard, as these are technical limitations on how App Store Connect sessions are handled.

### Separate Apple ID for CI

> Apple announced that as of February 27th 2019, it is enforcing 2FA on developer Apple ID's thus disabling the option to create a separate Apple ID specifically for CI without 2FA. If you somehow still have an Apple ID that does not have 2 Factor Authentication enabled, you can of course use that one to login via _fastlane_.

Reviewer 2: Shouldn't it be `... on developer Apple IDs, thus disabling ...`?

Author: yep:

```suggestion
Apple announced that as of February 27th 2019, it is enforcing 2FA on developer Apple IDs thus disabling the option to create a separate Apple ID specifically for CI without 2FA. If you somehow still have an Apple ID that does not have 2 Factor Authentication enabled, you can of course use that one to login via _fastlane_.
```

> Apple announced that as of February 27th 2019, it is enforcing 2FA on developer Apple ID's thus disabling the option to create a separate Apple ID specifically for CI without 2FA. If you somehow still have an Apple ID that does not have 2 Factor Authentication enabled, you can of course use that one to login via _fastlane_.

Reviewer 3: The document is using _2-factor auth_ rather than the shorthand _2FA_. We should probably be consistent with it. This one's on me since I added this phrase using _2FA_, sorry about that.

Reviewer 4: Great point. I think we should also link to the official support docs. I'm not sure if all people know what 2-factor and 2-step authentication is and what the difference (and requirements) are.
[Two-factor authentication for Apple ID
](https://support.apple.com/en-us/HT204915)
[Two-step verification for Apple ID
](https://support.apple.com/en-us/HT204152)

Author: (You can both just suggest changes to the PR here. When you add a comment to the line you want to change, the first icon above the text field is two pages with a `+ -` in it. Clicking it adds the current text into the comment box, and then you can edit it. When you save the comment, I can just "accept" your suggestion)

Author: Links were added, shorthand was introduced via brackets

- [Azure DevOps](/best-practices/continuous-integration/azure-devops/) (formerly known as: Visual Studio Team Services)
- [NeverCode](/best-practices/continuous-integration/nevercode/)

## Authentication with Apple services

Several Fastlane actions communicate with Apple services that need authentication. This can post several challenges on CI:

### Separate Apple ID for CI

> The easiest way to get _fastlane_ running on a CI system is to create a separate Apple ID that doesn't have 2-factor authentication enabled.

Reviewer: ```suggestion
The easiest way to get _fastlane_ running on a CI system is to create a separate Apple ID that 

  - doesn't have 2-factor authentication enabled.
  - doesn't have the Account Holder role
```

Creating a separate Apple ID allows you to limit the permission scope (limited access to only the apps and resources it needs), have a long, randomly generated password, and will make it much more convenient for you to set up CI using _fastlane_.

### Two-step or Two-factor auth

_fastlane_ fully supports [2-factor authentication (2FA)](https://support.apple.com/en-us/HT204915) (and legacy [2-step verification (2SV)](https://support.apple.com/en-us/HT204152)) for logging in to your Apple ID and Apple Developer account. 🌟

> Note: Apple announced that as of February 27th 2019, it is enforcing 2-factor authentication on developer Apple IDs with the "Account Owner" role.

Reviewer: ```suggestion
Note: [Apple announced that as of February 27th 2019](https://developer.apple.com/news/?id=02202019a), it is enforcing 2-factor authentication on developer Apple IDs with the "Account Holder" role.
```

The resulting session will be stored in `~/.fastlane/spaceship/[email]/cookie`, which should be valid for about one month.

#### Use of application specific passwords and `spaceauth`

When you can not enter the security code manually, as on a Continuous Integration system, you have to use other ways to log in:

##### Application specific passwords

If you want to upload builds to App Store Connect (actions `upload_to_app_store` and `deliver`) or TestFlight (actions `upload_to testflight`, `pilot` or `testflight`) from your CI machine, you need to generate an _application specific password_:

1. Visit [appleid.apple.com/account/manage](https://appleid.apple.com/account/manage)
1. Generate a new application specific password
1. Provide the application specific password using the environment variable `FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD`

This will supply the application specific password to iTMSTransporter, the tool used by those actions to perform the upload.

##### `spaceauth`

All other actions interacting with Apple's APIs do not accept application specific passwords.

> As your CI machine will not be able to prompt you for your two-factor authentication or two-step verification information, you need to generate a login session for Apple ID  in advance. You can get on your local machine this by running:

Reviewer: ```suggestion
As your CI machine will not be able to prompt you for your two-factor authentication or two-step verification information, you need to generate a login session for Apple ID in advance. You can get on your local machine this by running:
```

- [Azure DevOps](/best-practices/continuous-integration/azure-devops/) (formerly known as: Visual Studio Team Services)
- [NeverCode](/best-practices/continuous-integration/nevercode/)

## Authentication with Apple services

> Several Fastlane actions communicate with Apple services that need authentication. This can post several challenges on CI:

Author: ```suggestion
Several Fastlane actions communicate with Apple services that need authentication. This can pose several challenges on CI:
```

### Separate Apple ID for CI

The easiest way to get _fastlane_ running on a CI system is to create a separate Apple ID that 

>   - doesn't have 2-factor authentication enabled.

Author: ```suggestion
  - doesn't have 2-factor authentication enabled
```

})();
</script>

> The following tool and service specific content was moved:

Reviewer: ```suggestion
The following tool- and service-specific content was moved:
```
